### PR TITLE
Add a method to `Targets` to get the default level

### PR DESCRIPTION
**Pull request by connec**
## Motivation

This makes it possible to fully "override" some base `Targets` filter with another (e.g. user-supplied) filter. Without some way to obtain the default, only explicit targets can be overridden (via `IntoIter` and friends).

## Solution

We can add a method to `Targets` that filters the underlying `DirectiveSet` for the default directive. This works because `DirectiveSet::add` will replace directives with the same `target`/`field_names`, which is always `None`/`vec![]` for the directive added by `with_default` (and in fact we are only concerned with `target`, since no other `Targets` API allows adding directives with a `None` target).

Ideally the method would be named `default`, corresponding to `with_default`, however this conflicts with `Default::default` and so would be a breaking change (and harm ergonomics). `default_level` seemed a better name than `get_default`, since "getters" of this style are generally considered unidiomatic<sup>[citation needed]</sup>.

Example usage:

```rust
let mut filter = Targets::new().with_target("some_module", LevelFilter::INFO);

// imagine this came from `RUST_LOG` or similar
let override: Targets = "trace".parse().unwrap();

// merge the override
if let Some(default) = override.default_level() {
    filter = filter.with_default(default);
}
for (target, level) in override.iter() {
    filter = filter.with_target(target, level);
}
```

### CODE REVIEW

File: tracing-subscriber/src/filter/targets.rs [hunk 1]
```diff
@@ -277,6 +277,31 @@ impl Targets {
...
+    /// about the behaviour you can use `unwrap_or`:
+    ///
+    /// ```
+    /// use tracing_subscriber::filter::{LevelFilter, Targets};
+    ///
+    /// let filter = Targets::new();
+    /// let default_level = filter.default_level().unwrap_or(LevelFilter::OFF);
+    ///
+    /// assert_eq!(default_level, LevelFilter::OFF);
+    /// ```
```
Comment (hawkw) [inline tracing-subscriber/src/filter/targets.rs:296]: it might be nice to have an example that explicitly demonstrates when `None` is returned, and when `Some(LevelFilter::OFF)` is returned?

File: tracing-subscriber/src/filter/targets.rs [hunk 2]
```diff
@@ -277,6 +277,31 @@ impl Targets {
...
+    pub fn default_level(&self) -> Option<LevelFilter> {
+        self.0
+            .directives()
+            .into_iter()
+            .find(|d| d.target.is_none())
+            .map(|d| d.level)
```
Comment (hawkw) [inline tracing-subscriber/src/filter/targets.rs:302]: tiny style nit, take it or leave it: this could also be expressed using [`Iterator::find_map`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.find_map)
```suggestion
            .find_map(|d| d.target.is_none().then(d.level))
```

File: tracing-subscriber/src/filter/targets.rs [hunk 3]
```diff
@@ -277,6 +277,31 @@ impl Targets {
...
     }
 
+    /// Returns the explicitly set default level for this filter, if any.
+    ///
+    /// If a default level was set explicitly using [`with_default`](Self::with_default), that level
+    /// will be returned. Otherwise `None` will be returned.
+    ///
+    /// A return value of `None` is behaviourly equivalent to [`LevelFilter::OFF`], but
+    /// distinguishes between an explicitly set default level or the default one. If you only care
+    /// about the behaviour you can use `unwrap_or`:
```
Comment (hawkw) [inline tracing-subscriber/src/filter/targets.rs:287]: style nit, not a big deal: generally, we try to avoid using second-person ("you") language in API docs, can we reword this to avoid "you"? i know @davidbarsky cares about this...

File: tracing-subscriber/src/filter/targets.rs [hunk 4]
```diff
@@ -277,6 +277,31 @@ impl Targets {
         self
     }
 
+    /// Returns the explicitly set default level for this filter, if any.
+    ///
+    /// If a default level was set explicitly using [`with_default`](Self::with_default), that level
+    /// will be returned. Otherwise `None` will be returned.
```
Comment (hawkw) [inline tracing-subscriber/src/filter/targets.rs:283]: this will also return a `LevelFilter` if the `Targets` was parsed from a filter string that contains a level with no target. we should mention that here (and maybe also add an example showing what this method would return when the `Targets` is parsed from a string?

**Review by hawkw — CHANGES_REQUESTED** [review]
This definitely seems worth having, and I think the `default_level` naming is fine. I had some very minor suggestions about the API docs, but once those are addressed I'll be very happy to merge this!

Thank you!

File: tracing-subscriber/src/filter/targets.rs (hunk 2 above)
Comment (connec) [reply tracing-subscriber/src/filter/targets.rs]: Good catch, I did have that at some point along with `bool::then`. Think I got muddled up when checking the MSRV.

File: tracing-subscriber/src/filter/targets.rs (hunk 4 above)
Comment (connec) [reply tracing-subscriber/src/filter/targets.rs:283]: Good catch, that's actually the case I'm most interested in (being able to set the default from a `Targets` parsed from user input).

File: tracing-subscriber/src/filter/targets.rs (hunk 1 above)
Comment (connec) [reply tracing-subscriber/src/filter/targets.rs]: I've bulked out the examples, perhaps excessively, to cover `with_default`/`parse` for all of set/unset/set to `OFF`.

File: tracing-subscriber/src/filter/targets.rs (hunk 2 above)
Comment (connec) [reply tracing-subscriber/src/filter/targets.rs]: Oh, I remember now. It's because `bool::then` is only available since 1.50.0 and MSRV is 1.49.0. We can use `filter_map` with an `if`/`else` but that's a bit more verbose.

```rust
            .find_map(|d| {
                if d.target.is_none() {
                    Some(d.level)
                } else {
                    None
                }
            })
```

**Comment by connec** [comment]
Thanks for the fast review! I think I've responded to all your comments.

I did end up second-guessing the naming a little bit, since what we really return is a `LevelFilter` rather than a `Level` (and `LevelFilter` is already an `Option<Level>` under the hood). It does match up with the `level` argument for `with_default` though so perhaps it's fine (I'm not sure `default_level_filter()` is much better?).

**Comment by hawkw** [comment]
> I did end up second-guessing the naming a little bit, since what we really return is a `LevelFilter` rather than a `Level` (and `LevelFilter` is already an `Option<Level>` under the hood). It does match up with the `level` argument for `with_default` though so perhaps it's fine (I'm not sure `default_level_filter()` is much better?).

Personally, I think `default_level` is fine --- the return type should be pretty obvious just from looking at the function signature; I wouldn't worry about it too much. :)

File: tracing-subscriber/src/filter/targets.rs [hunk 5]
```diff
@@ -277,6 +277,59 @@ impl Targets {
         self
     }
 
+    /// Returns the default level for this filter, if set.
+    ///
```
Comment (hawkw) [inline tracing-subscriber/src/filter/targets.rs:281]: maybe worth adding something like:
```suggestion
    ///
    /// The default level is used to filter any spans or events with targets
    /// that do not match any of the configured set of prefixes.
    ///
```

File: tracing-subscriber/src/filter/targets.rs [hunk 6]
```diff
@@ -277,6 +277,59 @@ impl Targets {
         self
     }
 
+    /// Returns the default level for this filter, if set.
```
Comment (hawkw) [inline tracing-subscriber/src/filter/targets.rs:280]: tiny nit:
```suggestion
    /// Returns the default level for this filter, if one is set.
```

File: tracing-subscriber/src/filter/targets.rs [hunk 7]
```diff
@@ -277,6 +277,59 @@ impl Targets {
...
+    /// use tracing_subscriber::filter::Targets;
+    ///
+    /// let filter = Targets::new();
+    /// assert_eq!(filter.default_level(), None);
+    ///
+    /// let filter: Targets = "my_crate=info".parse().unwrap();
+    /// assert_eq!(filter.default_level(), None);
+    /// ```
+    ///
+    /// Note that an unset default level (`None`) behaves like `LevelFilter::OFF` when the filter is
```
Comment (hawkw) [inline tracing-subscriber/src/filter/targets.rs:310]: nit, take it or leave it: 
```suggestion
    /// Note that an unset default level (`None`) behaves like [`LevelFilter::OFF`] when the filter is
```

**Review by hawkw — APPROVED** [review]
this looks good to me, thank you again!

**Comment by hawkw** [comment]
I believe this also closes #1790, as it implements the solution discussed in https://github.com/tokio-rs/tracing/issues/1790#issuecomment-999739222.
